File: Introductory Problems/TowerOfHanoi/tower_of_hanoi.cpp
#include<bits/stdc++.h>
using namespace std;
#define ll long long int

vector<vector<int>> ans;
void move(int src, int dest){
	vector<int> temp;
	temp.push_back(src);
	temp.push_back(dest);
	ans.push_back(temp);
	return;
}
void towerOfHanoi(int n, int src, int dest, int helper){
	if(n == 1){
		move(src, dest);
		return;
	}
	towerOfHanoi(n-1, src, helper, dest);
	move(src, dest);
	towerOfHanoi(n-1, helper, dest, src);

}
void solve(){
	long n;
	cin >> n;
	towerOfHanoi(n, 1, 3, 2);
	cout<<ans.size()<<endl;
	for(int i=0;i<ans.size();i++){
		cout<<ans[i][0]<<" "<<ans[i][1]<<endl;
	}
	return;
	
}

int main(){
	long long testCases;
	testCases = 1;
	// cin>>testCases;
	while(testCases--){
		solve();
	}
	return 0;
}

File: Introductory Problems/TowerOfHanoi/towerOfHanoi.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class towerOfHanoi {
    static List<String> ans = new ArrayList();
    private static void move(int src, int dest){
        String temp = src + " " + dest;
        ans.add(temp);
    }
    private static void towerOfHanoiHelper(int n, int src, int dest, int helper){
        if(n == 1){
            move(src, dest);
            return;
        }
        towerOfHanoiHelper(n-1, src, helper, dest);
        move(src, dest);
        towerOfHanoiHelper(n-1, helper, dest, src);
    }
    private static void solve(){
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        towerOfHanoiHelper(n, 1, 3, 2);
        System.out.println(ans.size());
        for(int i = 0; i<ans.size(); i++){
            System.out.println(ans.get(i));
        }
        return;
    }
    public static void main(String args[]){
        solve();
    }
}
